articles/101_build_tool.md

```diff
@@ -0,0 +1,213 @@
+author: '[Dirk Thomas](https://github.com/dirk-thomas)'
+categories: Overview
+---
+
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+## Preface
+
+ROS is developed with a federated model.
+It consists of many separately developed and maintained packages.
+
+The "manual" approach to build a set of packages consists of building all packages in their topological order one by one.
+For each package the documentation usually describes what the dependencies are, how to setup the environment to build the package as well as how to setup the environment afterwards to use the package.
+From an efficiency point of view that manual process does not scale well.
+
+This article describes the steps from the current build tools used in the ROS ecosystem to a single universal build tool which automates the process and works for various different use cases.
+
+## Goal
+
+The goal of the build tool is to build a set of packages with a single invocation automating the process.
+It should work with ROS 1 packages as well as ROS 2 packages which provide the necessary information in their manifest files.
+It should also work with packages that do not provide manifest files themselves, given that the meta information is externally provided.
+This will allow the build tool to be utilized for non-ROS dependencies of ROS packages (e.g. Gazebo including its dependencies).
+
+### Out of Scope
+
+The build tool does not cover the steps necessary to fetch the sources of the to-be-built packages.
+This can be done with existing tools (e.g. [rosinstall_generator](http://wiki.ros.org/rosinstall_generator), [vcstool](https://github.com/dirk-thomas/vcstool)).
+
+The build tool also does not provide a mechanism to install any dependencies required to build the packages.
+In the ROS ecosystem [rosdep](http://wiki.ros.org/rosdep) can be used for this.
+
+The build tool also does not create binary packages (e.g. a Debian package).
+In the ROS ecosystem [bloom](http://wiki.ros.org/bloom) is used to generate the required metadata and then platform dependent tools like `dpkg-buildpackage` build binary packages.
+
+## Build Tool vs. Build System
+
+A build tool operates on a set of packages.
+It determines the dependency graph and invokes the package specific build tool of each package in topological order.
+The build tool itself should know as little as possible about the build system used for a specific package.
+Just enough in order to know how to setup the environment for it, invoke the build, and setup the environment to use the built package.
+
+The build system on the other hand operates on a single package.
+In the case a package uses e.g. `CMake` the responsibility of the tool comes down to invoke the common steps `cmake`, `make`, `make install` for this package.
+As another example for a package using `Autotools` the steps could look like `configure`, `make`, `make install`.
```

Comment:
I think it might be useful to create a table of things and how we would categorize them according to these descriptions. For example, you could list something like `CMake which calls catkin macros`, `Plain CMake`, `CMake that calls ament_cmake macros`, ```Python package that only has a `setup.py` ```, etc as a build system and then list `catkin_make`, `catkin_make_isolated`, `catkin build (catkin_tools)`, `ament build (ament_tools)`, etc as build tools. I think an explicit, if not exhaustive list, would help cement this idea for people first encountering this conceptual distinction.